### PR TITLE
Combine Dev.to API with Single Post UI

### LINKED ISSUE #216: Connect Single Blog Post UI with dev.to and route users to corresponding single blog post page
**What do we need to build or fix?**
We need to implement the logic to connect our single blog post UI with dev.to. This involves automatically pulling content data from dev.to and ensuring users are directed to the corresponding single post page on our Next.js website instead of being redirected to dev.to. This will help keep users on our website longer.

**Technical details**
_1. Set Up API Call to dev.to:_

- Extend the `getStaticProps` function in `pages/blog/posts/[id].js` to fetch detailed post content from dev.to using their API.
- Use the dev.to API to fetch the content for individual blog posts based on their id.

_2. Update Single Blog Post UI:_

- Utilize the fetched data in the BlogPostContainer component.
- Pass the necessary data (title, content, author, etc.) to the BlogPostContainer and PostContent.

_3. Routing Logic:_

- Ensure that clicking on a blog post link in the blog post list routes to the new single blog post page using dynamic routing.
- Update BlogPostsContainer to link to the new [id].js page.

**Approach suggestions**
 - https://dev.to/cherylli/displaying-devto-articles-on-your-website-using-nextjs-1bbk
 - Example of implementation for `pages/blog/posts/[id].js`

```
import BlogPostContainer from '@/components/blog/BlogPostContainer';
import { useRouter } from 'next/router';

const BlogPost = ({ post }) => {
  const router = useRouter();
  if (router.isFallback) {
    return <div>Loading...</div>;
  }
  return <BlogPostContainer post={post} />;
};

export default BlogPost;

export async function getStaticPaths() {
  const res = await fetch('https://dev.to/api/articles?username=wdp');
  const posts = await res.json();

  const paths = posts.map(post => ({
    params: { id: post.id.toString() },
  }));

  return { paths, fallback: true };
}

export async function getStaticProps({ params }) {
  const res = await fetch(`https://dev.to/api/articles/${params.id}`);
  const post = await res.json();
  const userSummary = post.user.summary;

  return {
    props: {
      post: {
        id: post.id,
        title: post.title,
        cover_image: post.cover_image,
        published_at: post.published_at,
        user: {
          name: post.user.name,
          summary: userSummary,
        },
        body_html: post.body_html,
      },
    },
    revalidate: blogRevalidate,
  };
}
```

**Deadline**
Please keep in mind that once you assign this task to yourself, you'll need to complete it in 15 days.

**Acceptance criteria**

- Test the routing logic to ensure users are correctly navigated to the single post pages on our website.
- Test the page and components in many screen sizes, you can use the Inspect tool for that.
- Please test if the new changes added to the components do not affect the other instances.
- Test the feature in many browsers, such as Chrome, Firefox, Edge, and Safari (MAC).
- If there are any build problems when submitting your PR, run `yarn build` locally to solve the issues and commit the changes.
- Update the CHANGELOG.md file.

### PR DESCRIPTION
**Pull request by Satoshi-Sh**
|                                    Web Dev Path                                    |
| :--------------------------------------------------------------------------------: |
| [**216**](https://github.com/MarianaSouza/web-dev-path/issues/issueNumber) |

close #216
#### Have you updated the CHANGELOG.md file? If not, please do it.
Yes.
#### What is this change?
- Replaced the blog post link of Dev.to with the internal link.
- Combined Dev.to API with Single post UI
- Modified styling issues accordingly(code block, youtube video, account without user summary...)
#### Were there any complications while making this change?
I didn't know how the getStaticPaths and getStaticProps functions work in Next.js. It was confusing at first. 
#### How to replicate the issue?
N.A.
#### If necessary, please describe how to test the new feature or fix.
Please go check individual blog post from the blog page. 

#### When should this be merged?
After reviews.

#### Screenshots
blog-post-demo2

## P.S. 
This part will be implemented after this PR?
Screenshot from 2024-06-21 13-52-36

### CODE REVIEW

**Comment by Satoshi-Sh** [comment]
I switched server-side rendering to client-side rendering as the app was built to hit Dev.to API rate limit.

There might be a way to avoid hitting the limit while using server-side rendering, but CSR works fine here.

**Comment by cherylli** [comment]
Had a good look at this today. I was not able to find any information about dev.to API limit
I have tried to set it to an extremely long delay and still getting that error (429)

What worked for me was to remove the part which fetch the user details (maybe they don't want you to use multiple endpoints too soon after another)

image

We could just have all the author's bio locally in `utils/about.js`, so we don't have to pull from dev.to which I think should be ok as we won't have random contributors. But we should still set a fallback.

I think we should use ISR, instead of client side rendering if possible, since the pages are crawlable which means better for SEO and also it's faster than client side rendering for the user

==========================

With the functionalities, there are a few things I noticed,
- I think clicking on the photo or anywhere in the card should also link to the single page, instead of just "read more". I keep trying to click on the image, so this can be an expected behaviour, thoughts?
- Maybe we should add "loading state" for the page while it's still loading?
- I noticed the linkedIn icon doesn't link to the authors linkedIn

**Comment by Satoshi-Sh** [comment]
>We could just have all the author's bio locally in utils/about.js, so we don't have to pull from dev.to which I think should be ok as we won't have random contributors. But we should still set a fallback.

It's a bit of work to hard coding but I think not many new people join and write a blog. I will take this path.

> I noticed the LinkedIn icon doesn't link to the author's linkedIn

Dev to account doesn't hold a LinkedIn account. That's why getting redirected to the WDP account as a fallback. I can use local data to put the individual account links.

**Comment by Satoshi-Sh** [comment]
## Updates 
At first, I was making a call to get user data from Dev.to API as post data does't include user summary.  @cherylli found that this call caused the error, and she advised me to copy the summary to our local member data and use them instead of making an extra API call.

Additionally, after this, I implemented a related/latest posts feature.

## Another Issue I found 
"Tips for collaborating with a new project codebase" has a styling issue and I'm unsure how to fix it. Need to update styling here `components/blog/PostContent/styles.js` somehow.

Screenshot from 2024-06-23 09-46-22

**Comment by cherylli** [comment]
I see the card has a pointer now but nothing happens when I clicked on it.

**Comment by Satoshi-Sh** [comment]
> I see the card has a pointer now but nothing happens when I clicked on it.

Are you talking about this?

> I think clicking on the photo or anywhere in the card should also link to the single page, instead of just "read more". I keep trying to click on the image, so this can be an expected behaviour, thoughts?

If so, I think this is out of the scope in this PR.

**Comment by cherylli** [comment]
This is what I mean. Is anyone else getting this? 
in the preview it's like this
image
on live it's like this
image
so when the cursor is pointer i'd expect it to be a clickable. The live version feels ok as it's not a pointer.

**Comment by cherylli** [comment]
I think the pointer issue is just the netlify preview, it seems fine locally. 🤔
I thought the preview should resemble the actual version

Anyway, everything looks good but I would add some spacing in between each "related post"

image

**Comment by Satoshi-Sh** [comment]
> Anyway, everything looks good but I would add some spacing in between each "related post"

Updated 

Screenshot from 2024-06-26 09-23-32

**Review by cherylli — APPROVED** [review]
looks good

File: CHANGELOG.md [hunk 1]
```diff
@@ -122,6 +122,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Added Kaz to 'about us'
 - Added theming to the project
 - Created single blog post page
+- Combined Dev.to API with single post UI
```
[the remark this answers is not in the document]
Comment (cherylli) [reply CHANGELOG.md:125]: oh wow I must have forgotten to post this this was like 3 reviews ago 😂

**Comment by Satoshi-Sh** [comment]
Thanks for the review and guidance!! @cherylli

File: utils/about.js [hunk 2]
```diff
@@ -33,6 +34,9 @@ export const whoWeAre = [
     image: '/images/klesta.webp',
     name: 'Klesta Luli',
     title: '',
+    devtoId: 897608,
```
Comment (mariana-caldas) [inline utils/about.js:37]: Good thinking about reusing the object!

**Review by mariana-caldas — APPROVED** [review]
Amazing job, @Satoshi-Sh ! 🚀 🥇

**Pull request merged**